core/server/common/src/main/java/alluxio/master/journal/AsyncJournalWriter.java

```diff
@@ -99,49 +162,193 @@ public long appendEntry(JournalEntry entry) {
+            // No more entries in the queue. Break write session.
             break;
           }
           mJournalWriter.write(entry);
           // Remove the head entry, after the entry was successfully written.
           mQueue.poll();
-          writeCounter = mWriteCounter.incrementAndGet();
+          mWriteCounter++;
 
-          if (writeCounter >= targetCounter) {
-            if ((System.nanoTime() - startTime) >= mFlushBatchTimeNs) {
-              // This thread has been writing to the journal for enough time. Break out of the
-              // infinite for-loop.
-              break;
+          if (((System.nanoTime() - startTime) >= mFlushBatchTimeNs) && !mStopFlushing) {
+            // This thread has been writing to the journal for enough time. Break out of the
+            // infinite while-loop.
+            break;
+          }
+        }
+
+        // Either written new entries or previous flush had been failed.
+        if (mFlushCounter.get() < mWriteCounter) {
+          mJournalWriter.flush();
+          mFlushCounter.set(mWriteCounter);
+        }
+
+        /**
+         * Notify tickets that have been served to wake up.
+         */
+        try (LockResource lr = new LockResource(mTicketLock)) {
+          ListIterator<FlushTicket> ticketIterator = mTicketList.listIterator();
+          while (ticketIterator.hasNext()) {
+            FlushTicket ticket = ticketIterator.next();
+            if (ticket.getTargetCounter() <= mFlushCounter.get()) {
+              ticket.setCompleted();
+              ticketIterator.remove();
             }
           }
         }
+      } catch (IOException | JournalClosedException exc) {
+        /**
+         * Release only tickets that have been flushed. Fail the rest.
+         */
+        try (LockResource lr = new LockResource(mTicketLock)) {
+          ListIterator<FlushTicket> ticketIterator = mTicketList.listIterator();
+          while (ticketIterator.hasNext()) {
+            FlushTicket ticket = ticketIterator.next();
+            ticketIterator.remove();
+            if (ticket.getTargetCounter() <= mFlushCounter.get()) {
+              ticket.setCompleted();
+            } else {
+              ticket.setError(exc);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Submits a ticket to flush thread and waits until ticket is served.
+   *
+   * If the specified counter is already flushed, this is essentially a no-op.
+   *
+   * @param targetCounter the counter to flush
+   */
+  @SuppressWarnings("Duplicates")
+  public void flush(final long targetCounter) throws IOException, JournalClosedException {
+    // Return if flushed.
+    if (targetCounter <= mFlushCounter.get()) {
+      return;
+    }
+
+    /**
+     * Submit the ticket for flush thread to process.
+     */
+    FlushTicket ticket = new FlushTicket(targetCounter);
+    try (LockResource lr = new LockResource(mTicketLock)) {
```

Comment:
could many flushing requests overwhelm the flushing threads? cause us to be way behind?

Reply:
Request count should not effect flushing performance. There may be a contention over locking on ticket list, but I wouldn't expect that to effect flush latency compared old model. 

We had discussed using concurrent list, however that could cause tickets to be overlooked until next flush tick, increasing latency on the rpc side.

This is definitely something to keep eye on and try to optimize if adverse effects are seen. However I didn't see any problems with the scale of FSMeta tests at least.

Reply:
@calvinjia after the discussion above, I reverted the change to use synchronized block and  now using ReentrantLock again. Because non-fairness of synchronized block may cause flush thread to starve under heavy load.